hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE LambdaCase                #-}
+{-# LANGUAGE OverloadedStrings         #-}
+module Ide.Plugin.Error (
+      -- * Plugin Error Handling API
+    PluginError(..),
+    runExceptT,
+    withExceptT,
+    pluginResponseM,
+    hoistExceptT,
+    hoistMaybeT,
+    handleMaybe,
+    handleMaybeM,
+    getNormalizedFilePathE,
+) where
+
+import           Control.Monad.Extra         (maybeM)
+import           Control.Monad.IO.Class      (MonadIO, liftIO)
+import           Control.Monad.Trans.Class   (lift)
+import           Control.Monad.Trans.Except  (ExceptT (..), mapExceptT,
+                                              runExceptT, throwE, withExceptT)
+import           Control.Monad.Trans.Maybe   (MaybeT, mapMaybeT)
+import qualified Data.Text                   as T
+import           Language.LSP.Protocol.Types
+import           Prettyprinter
+
+-- ----------------------------------------------------------------------------
+-- Plugin Error wrapping
+-- ----------------------------------------------------------------------------
+
+pluginResponseM :: Monad m => (t -> m (Either a b)) -> ExceptT t m b -> m (Either a b)
+pluginResponseM handler act =
+    runExceptT act >>= \case
+        Right r  -> pure $ Right r
+        Left err -> handler err
+
+-- See Note [PluginError]
+data PluginError
+  = -- |PluginInternalError should be used if something has gone horribly wrong.
+    -- All uncaught exceptions will be caught and converted to this error.
+    --
+    -- This error will be logged individually with Error, and will be converted
+    -- into an InternalError response code. It takes the highest precedence (1)
+    -- in being returned as a response to the client.
+    PluginInternalError T.Text
+    -- |PluginInvalidParams should be used if the parameters of the request are
+    -- invalid. This error means that there is a bug in the client's code
+    -- (otherwise they wouldn't be sending you requests with invalid
+    -- parameters).
+    --
+    -- This error will be logged individually with Warning, and will be
+    -- converted into a InvalidParams response code. It takes medium precedence
+    --  (2)in being returned as a response to the client.
+  | PluginInvalidParams T.Text
+    -- |PluginInvalidRequest should be used if the request is invalid. This
+    -- error means that there is a bug in the client's code (otherwise they
+    -- wouldn't be sending you an invalid request).
```

Comment:
what's the difference to the previous one?

Reply:
They correlate to the ErrorCodes of the same name. I would say InvalidParams is specifically for parameters, and that InvalidRequest is broader. Mostly, I defined both because both were used in the code base.

Reply:
Yeah, looking at the JSON-RPC spec again, its says that ParseError is if the JSON itself is invalid! So we shouldn't be using it at all. Probably we should indeed be using the LSP-specific codes like `RequestFailed` and so on.